### PR TITLE
env: set XDG_DATA_DIRS for wayland et.al.

### PR DESCRIPTION
**Pull request by sergiusens**
Currently XDG_DATA_DIRS is only setup for an X session, there
is currently no specific location to set this up for anything
using the wayland protocol. The profile.d mechanism seems like
the next best choice.

LP: #1681547

Signed-off-by: Sergio Schvezov <sergio.schvezov@ubuntu.com>

### CODE REVIEW

File: etc/profile.d/snapd.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# set XDG_DATA_DIRS for snapd provisioned desktop files.
+if [ "${XDG_DATA_DIRS#*snapd}" = "${XDG_DATA_DIRS}" ]; then
```
Comment (zyga) [inline etc/profile.d/snapd.sh]: Is this valid sh (not bash) syntax?

**Comment by morphis** [comment]
This very much looks like a copy of https://github.com/snapcore/snapd/blob/master/etc/X11/Xsession.d/65snappy 

Can't we link both somehow?

File: etc/profile.d/snapd.sh (hunk 1 above)
Comment (ogra1) [inline etc/profile.d/snapd.sh]: yes, thats appropriate to split variable content

**Comment by chipaca** [comment]
What's the status of this PR?

**Comment by chipaca** [comment]
Unless anybody objects I'm going to take over this PR, and unify all this into a single file. Ubuntu and Debian would ship it as `/etc/profile.d/apps-bin-path.sh` (because that's what we ship today, it's picked up by Xsession, and it's a conffile, and renaming conf files is AFAIK LOLNO territory), and it'd replace what's currently in snapd.spec and arch/snapd.sh (although tbh it might look alarmingly light the latter)

**Comment by chipaca** [comment]
are you sure profile.d isn't picked up by X? I moved 65snappy to /etc/profile.d/snapd.sh and started unity, xfce and jwm (heh) and in all cases `ps eww` of the toplevel process showed XDG_DATA_DIRS having snapd in it

**Comment by chipaca** [comment]
I mean I do this:

    ~$ ps fx
      PID TTY      STAT   TIME COMMAND
    16758 ?        Ss     0:00 /sbin/upstart --user
    16845 ?        S      0:00  \_ upstart-udev-bridge --daemon --user
    16852 ?        Ss     0:04  \_ dbus-daemon --fork --session --address=unix:abstract=/tmp/dbus-qFtZh3OQOM
    16885 ?        S      0:00  \_ upstart-dbus-bridge --daemon --system --user --bus-name system
    ...
so then

    ~$ ps eww 16758 | egrep -o 'XDG_DATA_DIRS=[^ ]*'
    XDG_DATA_DIRS=/usr/share/ubuntu:/usr/share/gnome:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop

**Comment by pedronis** [comment]
@chipaca is working on adjusting this

File: packaging/ubuntu-14.04/rules [hunk 2]
```diff
@@ -117,6 +117,7 @@ endif
 	# XXX: hacky
 	$(MAKE) -C cmd distclean || true
 	$(MAKE) -C data/systemd clean
+	$(MAKE) -C data/env clean
```
Comment (mvo5) [inline packaging/ubuntu-14.04/rules]: Hm, could we have a single `make -C data clean` here?

File: packaging/ubuntu-14.04/rules [hunk 3]
```diff
@@ -130,6 +131,9 @@ override_dh_auto_build:
 	cd cmd && ( ./configure --prefix=/usr --libexecdir=/usr/lib/snapd $(VENDOR_ARGS))
 	$(MAKE) -C cmd all
 
+	# Generate the env snippet
+	$(MAKE) -C data/env all
```
Comment (mvo5) [inline packaging/ubuntu-14.04/rules]: Could we have a single "$(MAKE) -C data all" here?

File: packaging/ubuntu-16.04/rules [hunk 4]
```diff
@@ -106,6 +106,7 @@ endif
 	# XXX: hacky
 	$(MAKE) -C cmd distclean || true
 	$(MAKE) -C data/systemd clean
```
Comment (mvo5) [inline packaging/ubuntu-16.04/rules]: Same question as for the 14.04 branch

**Comment by mvo5** [comment]
Fwiw, both lightdm and gdm3 load /etc/profile which in turn will load /etc/profile.d (in gdm3-3.24.1/data/Xsession.in and lightdm-1.22.0/debian/lightdm-session).

File: packaging/ubuntu-14.04/rules (hunk 3 above)
Comment (chipaca) [reply packaging/ubuntu-14.04/rules]: 14.04 wasn't making dbus. Is that a bug?

Comment (mvo5) [reply packaging/ubuntu-14.04/rules]: Probably :/

File: packaging/ubuntu-14.04/control [hunk 5]
```diff
@@ -114,10 +114,10 @@ Description: Transitional package for snapd
 
 Package: ubuntu-core-launcher
 Architecture: any
-Depends: snap-confine (= ${binary:Version}), ${misc:Depends}
-Section: oldlibs 
+Depends: snapd (= ${binary:Version}), ${misc:Depends}
+Section: oldlibs
 Pre-Depends: dpkg (>= 1.15.7.2)
-Description: Transitional package for snap-confine
+Description: Transitional package for snapd
```
Comment (mvo5) [inline packaging/ubuntu-14.04/control]: :heart:

File: packaging/ubuntu-14.04/snapd.autoimport.service [hunk 6]
```diff
@@ -1,10 +0,0 @@
-[Unit]
```
Comment (mvo5) [inline packaging/ubuntu-14.04/snapd.autoimport.service]: :heart: :heart:

**Comment by niemeyer** [comment]
@chipaca Thanks for working on this. The current error looks related:
```
+ echo 'Ensure sudo works without a password inside classic'
Ensure sudo works without a password inside classic
++ tr -d '\r'
++ classic sudo id -u
+ '[' '/etc/profile: line 23: .: /etc/profile.d/apps-bin-path.sh: is a directory
0' '!=' 0 ']'
+ echo 'sudo inside classic did not work as expected'
sudo inside classic did not work as expected
```

**Comment by chipaca** [comment]
@niemeyer yes, looking into why we're getting a directory there. Probably some silly error on my end.

**Comment by Swoorup** [comment]
This issue still exist on 18.04 when using wayland. `snap` applications are not listed under applications and cannot issue snap app commands on the terminal. I am using zsh. However this works fine under X11.

For some reason I can't log in to the launchpad site. So I have posted the findings below:

```bash
➜  ~ snap --version
snap    2.32.5+18.04
snapd   2.32.5+18.04
series  16
ubuntu  18.04
kernel  4.15.0-15-generic
➜  ~ uname -a      
Linux swoorup-W35xSS-370SS 4.15.0-15-generic #16-Ubuntu SMP Wed Apr 4 13:58:14 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
➜  ~ lsb_release -r
Release:	18.04
➜  ~ env | grep DATA
XDG_DATA_DIRS=/home/swoorup/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/
➜  ~ firefox
zsh: command not found: firefox
➜  ~ snap list
Name                     Version                  Rev   Tracking  Developer      Notes
atom                     1.26.0                   148   stable    snapcrafters   classic
core                     16-2.32.3                4407  stable    canonical      core
discord                  0.0.4                    52    stable    snapcrafters   -
firefox                  59.0.2-1                 71    stable    mozilla        -
intellij-idea-community  2018.1.1                 43    stable    jetbrains      classic
mailspring               1.2.1                    209   stable    foundry376     -
pycharm-community        2018.1.1                 58    stable    jetbrains      classic
spotify                  1.0.77.338.g758ebd78-41  13    stable    spotify        -
telegram-desktop         1.2.17                   142   stable    3v1n0          -
tusk                     0.10.1                   15    stable    klauscfhq      -
vlc                      3.0.1-4-g14a4897         190   stable    videolan       -
vscode                   1.20.1-1518535978        27    stable    flexiondotorg  classic
webstorm                 2018.1.1                 22    stable    jetbrains      classic
```

Let me know if you need additional information.

**Comment by stucki** [comment]
@Swoorup I confirm that the issue still occours with Ubuntu 18.04, zsh and snapd.

The root of the problem seems to be that the script in /etc/profile.d/apps-bin-path.sh, which takes care of proper setup, is sourced when starting an X11 session but not when starting a Wayland session (at least not in combination with zsh).

see also https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1640514

**Pull request merged**